pkg/internal/testing/controlplane/apiserver.go

```diff
@@ -0,0 +1,452 @@
+		return err
+	}
+
+	// NB(directxman12): insecure port is a mess:
+	// - 1.19 and below have the `--insecure-port` flag, and require it to be set to zero to
+	//   disable it, otherwise the default will be used and we'll conflict.
+	// - 1.20 requires the flag to be unset or set to zero, and yells at you if you configure it
+	// - 1.24 won't have the flag at all...
+	//
+	// In an effort to automatically do the right thing during this mess, we do feature discovery
+	// on the flags, and hope that we've "parsed" them properly.
+	//
+	// TODO(directxman12): once we support 1.20 as the min version (might be when 1.24 comes out,
+	// might be around 1.25 or 1.26), remove this logic and the corresponding line in API server's
+	// default args.
+	if err := s.discoverFlags(); err != nil {
+		return err
+	}
+
+	s.processState.Args, err = process.TemplateAndArguments(s.Args, s.Configure(), process.TemplateDefaults{
+		Data:     s,
+		Defaults: s.defaultArgs(),
+		// as per kubernetes-sigs/controller-runtime#641, we need this (we
+		// probably need other stuff too, but this is the only thing that was
+		// previously considered a "minimal default")
+		MinimalDefaults: map[string][]string{
+			"service-cluster-ip-range": []string{"10.0.0.0/24"},
+		},
+	})
+
+	return err
+}
+
+// discoverFlags checks for certain flags that *must* be set in certain
+// versions, and *must not* be set in others.
+func (s *APIServer) discoverFlags() error {
+	// Present: <1.24, Absent: >= 1.24
+	present, err := s.processState.CheckFlag("insecure-port")
+	if err != nil {
+		return err
+	}
+
+	if !present {
+		s.Configure().Disable("insecure-port")
+	}
+
+	return nil
+}
+
+func (s *APIServer) defaultArgs() map[string][]string {
+	args := map[string][]string{
+		"advertise-address":        []string{"127.0.0.1"},
+		"service-cluster-ip-range": []string{"10.0.0.0/24"},
+		"allow-privileged":         []string{"true"},
+		// we're keeping this disabled because if enabled, default SA is
+		// missing which would force all tests to create one in normal
+		// apiserver operation this SA is created by controller, but that is
+		// not run in integration environment
+		"disable-admission-plugins": []string{"ServiceAccount"},
+		"authorization-mode":        []string{"RBAC"},
+		"secure-port":               []string{s.SecureServing.Port},
+		// NB(directxman12): previously we didn't set the bind address for the secure
+		// port.  It *shouldn't* make a difference unless people are doing something really
+		// funky, but if you start to get bug reports look here ;-)
+		"bind-address": []string{s.SecureServing.Address},
+
+		// required on 1.20+, fine to leave on for <1.20
+		"service-account-issuer":           []string{s.SecureServing.URL("https", "/").String()},
+		"service-account-key-file":         []string{filepath.Join(s.CertDir, saCertFile)},
+		"service-account-signing-key-file": []string{filepath.Join(s.CertDir, saKeyFile)},
```

Comment:
Why does the Apiserver needs this if the CM creates the SAs?

Reply:
You know, I'm not sure, but it does :woman_shrugging: I'll look it up.